samza-azure/src/main/java/org/apache/samza/checkpoint/azure/AzureCheckpointManager.java

```diff
@@ -0,0 +1,248 @@
+ * The table entities take the following form:
+ *
+ * +-----------------+---------------------+-------------------+
+ * |                 |     Serialized      |                   |
+ * |   TaskName      |     JSON SSP        |     Offset        |
+ * |                 |                     |                   |
+ * +-----------------+---------------------+-------------------+
+ *
+ *  Each entity have a partitionKey set as the TaskName and the rowKey set as the SSP.
+ */
+public class AzureCheckpointManager implements CheckpointManager {
+  private static final Logger LOG = LoggerFactory.getLogger(AzureCheckpointManager.class.getName());
+  private static final String PARTITION_KEY = "PartitionKey";
+
+  public static final int MAX_WRITE_BATCH_SIZE = 100;
+  public static final String CHECKPOINT_MANAGER_TABLE_NAME = "SamzaTaskCheckpoints";
+  public static final String SYSTEM_PROP_NAME = "system";
+  public static final String STREAM_PROP_NAME = "stream";
+  public static final String PARTITION_PROP_NAME = "partition";
+
+  private final String storageConnectionString;
+  private final AzureClient azureClient;
+  private CloudTable cloudTable;
+
+  private final Set<TaskName> taskNames = new HashSet<>();
+  private final JsonSerdeV2<Map<String, String>> jsonSerde = new JsonSerdeV2<>();
+
+  AzureCheckpointManager(AzureConfig azureConfig) {
+    storageConnectionString = azureConfig.getAzureConnectionString();
+    azureClient = new AzureClient(storageConnectionString);
+  }
+
+  @Override
+  public void start() {
+    try {
+      // Create the table if it doesn't exist.
+      cloudTable = azureClient.getTableClient().getTableReference(CHECKPOINT_MANAGER_TABLE_NAME);
+      cloudTable.createIfNotExists();
+
+    } catch (URISyntaxException e) {
+      LOG.error("Connection string {} specifies an invalid URI while creating checkpoint table.",
+              storageConnectionString);
+      throw new AzureException(e);
+
+    } catch (StorageException e) {
+      LOG.error("Azure Storage failed when creating checkpoint table", e);
+      throw new AzureException(e);
+    }
+  }
+
+  @Override
+  public void register(TaskName taskName) {
+    taskNames.add(taskName);
+  }
+
+  @Override
+  public void writeCheckpoint(TaskName taskName, Checkpoint checkpoint) {
+    if (!taskNames.contains(taskName)) {
+      throw new SamzaException("writing checkpoint of unregistered task");
+    }
+
+    TableBatchOperation batchOperation = new TableBatchOperation();
+    checkpoint.getOffsets().forEach((ssp, offset) -> {
+        // Create table entity
+        TaskCheckpointEntity taskCheckpoint = new TaskCheckpointEntity(taskName.toString(),
+                serializeSystemStreamPartition(ssp));
+        taskCheckpoint.setOffset(offset);
+
+        // Add to batch operation
+        batchOperation.insertOrReplace(taskCheckpoint);
+
+        // Execute when batch reaches capacity
+        if (batchOperation.size() >= MAX_WRITE_BATCH_SIZE) {
+          try {
```

Comment:
consider consolidating this and the block in L:131-138

```
if (batchOperation.size() >= MAX_WRITE_BATCH_SIZE || !iterator.hasNext()) {
  table.execute(batchOperation)
}
```
This will get rid off the special redundant code to process the last batch